### PR TITLE
Allow strings as intent callout response. 

### PR DESCRIPTION
**Pull request by ShivamPokhriyal**
https://github.com/dimagi/commcare-android/issues/2321

Product Note: Allow arbitrary named strings as intent callout response.

### CODE REVIEW

File: app/src/org/commcare/android/javarosa/IntentCallout.java [hunk 1]
```diff
@@ -207,28 +213,44 @@ private boolean processOdkResponse(Intent intent, TreeReference intentQuestionRe
...
+                if (hasExtra) {
+                    return true;
                 }
             }
         }
         return (result != null);
     }
 
+    private void setOdkResponseValue(TreeReference intentQuestionRef, File destination, String responseValue, Vector<TreeReference> responseRefs) {
+        if (responseValue == null) {
```
Comment (shubham1g5) [inline app/src/org/commcare/android/javarosa/IntentCallout.java:245]: I think a better approach would be just to skip the call to `processResponseItem` when `responseValue` is null.

**Review by shubham1g5 — CHANGES_REQUESTED** [review]
One minor suggestion but otherwise we should add the respective tests for this behaviour in `IntentCalloutTests`

File: app/src/org/commcare/android/javarosa/IntentCallout.java (hunk 1 above)
Comment (ShivamPokhriyal) [reply app/src/org/commcare/android/javarosa/IntentCallout.java:245]: The earlier code uses empty value when `responseValue` is null. If that needs to be changed, then perhaps we can `return` here when the value is null?

Comment (shubham1g5) [reply app/src/org/commcare/android/javarosa/IntentCallout.java:245]: Ahh didn't realize it's like this before, no need for a change in behaviour in that case.

File: app/src/org/commcare/views/widgets/BarcodeWidget.java [hunk 2]
```diff
@@ -31,7 +31,7 @@ public BarcodeWidget(Context context, FormEntryPrompt prompt, PendingCalloutInte
     }
 
     @Override
-    protected void performCallout() {
+    public void performCallout() {
```
Comment (shubham1g5) [inline app/src/org/commcare/views/widgets/BarcodeWidget.java:34]: we should avoid changing method visibility for tests as much as possible. Instead you can just trigger the UI action that results in this being called out.

**Review by shubham1g5 — CHANGES_REQUESTED** [review]

File: app/src/org/commcare/views/widgets/BarcodeWidget.java (hunk 2 above)
Comment (ShivamPokhriyal) [reply app/src/org/commcare/views/widgets/BarcodeWidget.java:34]: makes sense. Will do the changes.

**Review by shubham1g5 — APPROVED** [review]
looks great.

**Comment by ShivamPokhriyal** [comment]
@damagatchi retest this please

**Comment by ShivamPokhriyal** [comment]
@damagatchi retest this please

**Comment by ShivamPokhriyal** [comment]
@damagatchi retest this please

**Comment by dimagimon** [comment]
[33% (0.0%) vs master 33%](https://jenkins.dimagi.com/job/commcare-android-pr-job-builder/462236/)

**Pull request merged**